builtin/k8s/status_report_util.go

```diff
@@ -0,0 +1,81 @@
+package k8s
+
+import (
+	corev1 "k8s.io/api/core/v1"
+
+	sdk "github.com/hashicorp/waypoint-plugin-sdk/proto/gen"
+)
+
+// Translate a K8S container status into a Waypoint Health Status
+func containerStatusToHealth(
+	containerStatus corev1.ContainerStatus,
+) *sdk.StatusReport_Resource {
+	var resourceHealth sdk.StatusReport_Resource
+	resourceHealth.Name = containerStatus.Name
+
+	// ContainerStatus.State is a struct of possible container states. If one
+	// is set, that is the current state of the container
+	if containerStatus.State.Running != nil || containerStatus.Ready == true {
+		resourceHealth.Health = sdk.StatusReport_READY
+		resourceHealth.HealthMessage = "container is reporting running" // no message defined by k8s api
+	} else if containerStatus.State.Waiting != nil {
+		resourceHealth.Health = sdk.StatusReport_PARTIAL
+		resourceHealth.HealthMessage = containerStatus.State.Waiting.Message
+	} else if containerStatus.State.Terminated != nil {
+		resourceHealth.Health = sdk.StatusReport_DOWN
+		resourceHealth.HealthMessage = containerStatus.State.Terminated.Message
+	} else {
+		resourceHealth.Health = sdk.StatusReport_UNKNOWN
+		resourceHealth.HealthMessage = "container health could not be determined"
+	}
+
+	return &resourceHealth
+}
+
+// Translate a Pod Phase into a Waypoint Health Status
+func podPhaseToHealth(
+	phase corev1.PodPhase,
+) sdk.StatusReport_Health {
+	var healthResult sdk.StatusReport_Health
+
+	switch phase {
+	case corev1.PodPending:
+		healthResult = sdk.StatusReport_ALIVE
+	case corev1.PodRunning:
+		healthResult = sdk.StatusReport_READY
+	case corev1.PodSucceeded:
+		healthResult = sdk.StatusReport_PARTIAL
+	case corev1.PodFailed:
+		healthResult = sdk.StatusReport_DOWN
+	case corev1.PodUnknown:
+		healthResult = sdk.StatusReport_UNKNOWN
+	default:
+		healthResult = sdk.StatusReport_UNKNOWN
+	}
+
+	return healthResult
+}
+
+// Take a list of Pods and build a Waypoint Status Report based on their reported health
+func buildStatusReport(
+	podList *corev1.PodList,
+) sdk.StatusReport {
+	var result sdk.StatusReport
+	result.External = true
+	resources := make([]*sdk.StatusReport_Resource, len(podList.Items))
+
+	// Report on most recently observed status of a deployments pod
+	for _, pod := range podList.Items {
+		// Overall Pod Health
+		podStatus := pod.Status
+		result.HealthMessage = podStatus.Message
+		result.Health = podPhaseToHealth(podStatus.Phase)
```

Comment:
It looks like `result` will carry the Health and HealthMessage of whatever the last pod was in your slice. I'm not sure exactly what the StatusReport adjectives mean, but I think you could declare the whole `result` as:

`StatusReport_READY` if all pods are ready

`statusReport_ALIVE` if all pods are in alive or ready, but not all are ready

`statusReport_DOWN` if all pods are down

`statusReport_UNKNOWN` if all pods are unknown? any pods? not sure no this one.

`statusReport_PARTIAL` for any mix that includes a down or an unknown?

Reply:
Oops, good catch!